api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java

```diff
@@ -90,11 +90,23 @@
      * Note that this property only gets evaluated during ConfigSource discovery.
      *
      * The default ordinals for the default config sources:
-     * <ol>
-     *  <li>System properties (default ordinal=400)</li>
-     *  <li>Environment properties (default ordinal=300)</li>
-     *  <li>/META-INF/microprofile-config.properties (default ordinal=100)</li>
-     * </ol>
+     <ol>
+     * <li>System properties (ordinal=400)</li>
+     * <li>Environment properties (ordinal=300)
+     *    <p>Some operating systems allow only alphabetic characters or an underscore(_), in environment variables.
+     *    Other characters such as ., /, etc may be disallowed.
+     *    In order to set a value for a config property that has a name containing such disallowed characters from an environment variable,
+     *    the following rules are used.
+     *    This ConfigSource searches for potentially 3 environment variables with a given property name (e.g. {@code "com.ACME.size"}):</p>
+     *        <ol>
+     *            <li>Exact match (i.e. {@code "com.ACME.size"})</li>
+     *            <li>Replace all illegal characters by '_' (i.e. {@code "com_ACME_size"})</li>
```

Comment:
I don't think mentioning "illegal characters" from the specification is correct.

How would you be able to determine which characters are illegal (from the OS env perspective) in Java to replace them.

I find the wording too open-ended and would prefer to constrain and make explicit which characters will be replaced.

Either we provide a fixed set (`.` is a must, we can also argue about `-` and `/`).

Another possibility is to specify that there can be a special env variable whose value is the list of characters that must be replaced. For example

```
MP_CONFIG_REPLACED_CHARS=".-/😀@"
```

The env config source will then replace any chars from that list by `_`.
By default, the value of MP_CONFIG_REPLACED_CHARS is `.-/`

Reply:
ah, my review was submitted after you merge the PR so I couldn't vote.
I am `-1` for this change

Reply:
as I mentioned on the call, I will clarify the meaning of illegal characters - basically any characters not alphabetic nor _ are viewed as illegal characters from environment variable name point of view.

Reply:
@Emily-Jiang + 1 (although I suppose you mean alphanumeric rather that alphabetic)

Reply:
@jmesnil yes, you were exactly right. I clarified via https://github.com/eclipse/microprofile-config/pull/327/files